src/Features/CSharp/Portable/ExtractLocalMethod/CSharpExtractLocalMethodCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,83 @@
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.ExtractLocalMethod
+{
+    [ExportCodeRefactoringProvider(LanguageNames.CSharp,
+        Name = PredefinedCodeRefactoringProviderNames.ExtractLocalMethod), Shared]
+    internal class CSharpExtractLocalMethodCodeRefactoringProvider : ExtractMethodCodeRefactoringProvider
+    {
+        [ImportingConstructor]
+        public CSharpExtractLocalMethodCodeRefactoringProvider()
+        {
+        }
+
+        protected override async Task<(CodeAction action, string methodBlock)> GetCodeActionAsync(
+            Document document,
+            TextSpan textSpan,
+            CancellationToken cancellationToken)
+        {
+            var documentOptions = await document.GetOptionsAsync(cancellationToken).ConfigureAwait(false);
+            var preferStatic = documentOptions.GetOption(CSharpCodeStyleOptions.PreferStaticLocalFunction).Value;
+
+            var result = await ExtractMethodService.ExtractMethodAsync(
+                document,
+                textSpan,
+                extractLocalMethod: true,
+                cancellationToken: cancellationToken).ConfigureAwait(false);
+            Contract.ThrowIfNull(result);
+
+            if (result.Succeeded || result.SucceededWithSuggestion)
+            {
+                // We insert an empty line between the generated local method and the previous statements if there is not one already.
+                var (resultDocument, resultMethodDeclarationNode, resultInvocationNameToken) = await InsertNewLineBeforeLocalMethodIfNecessaryAsync(result, cancellationToken).ConfigureAwait(false);
+
+                var description = FeaturesResources.Extract_Local_Method;
+
+                var codeAction = new MyCodeAction(description, c => AddRenameAnnotationAsync(resultDocument, resultInvocationNameToken, c));
+                var methodBlock = resultMethodDeclarationNode;
+
+                return (codeAction, methodBlock.ToString());
+            }
+
+            return default;
+        }
+
+        private async Task<(Document resultDocument, SyntaxNode resultMethodDeclarationNode, SyntaxToken resultInvocationNameToken)> InsertNewLineBeforeLocalMethodIfNecessaryAsync(
+            ExtractMethodResult result,
+            CancellationToken cancellationToken)
+        {
+            var resultDocument = result.Document;
+            var resultMethodDeclarationNode = result.MethodDeclarationNode;
+            var resultInvocationNameToken = result.InvocationNameToken;
+
+            // Checking to see if there is already an empty line before the local method declaration.
+            var leadingTrivia = result.MethodDeclarationNode.GetLeadingTrivia();
+            if (!leadingTrivia.Any(t => t.IsKind(SyntaxKind.EndOfLineTrivia)))
+            {
+                resultMethodDeclarationNode = result.MethodDeclarationNode.WithPrependedLeadingTrivia(SpecializedCollections.SingletonEnumerable(SyntaxFactory.CarriageReturnLineFeed));
```

Comment:
I wasn't exactly sure either. I ended up putting it here since the method is only necessary for local functions and is C# specific. I think otherwise it would have to go somewhere in the method extractor and probably needs both C#/VB implementations...